command/monitor/monitor.go

```diff
@@ -0,0 +1,172 @@
+package monitor
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+// Monitor provides a mechanism to stream logs using go-hclog
+// InterceptLogger and SinkAdapter. It allows streaming of logs
+// at a different log level than what is set on the logger.
+type Monitor interface {
+	// Start returns a channel of log messages which are sent
+	// ever time a log message occurs
+	Start() <-chan []byte
+
+	// Stop de-registers the sink from the InterceptLogger
+	// and closes the log channels
+	Stop()
+}
+
+// monitor implements the Monitor interface
+type monitor struct {
+	// protects droppedCount and logCh
+	sync.Mutex
+
+	sink log.SinkAdapter
+
+	// logger is the logger we will be monitoring
+	logger log.InterceptLogger
+
+	// logCh is a buffered chan where we send logs when streaming
+	logCh chan []byte
+
+	// doneCh coordinates the shutdown of logCh
+	doneCh chan struct{}
+
+	// droppedCount is the current count of messages
+	// that were dropped from the logCh buffer.
+	// only access under lock
+	droppedCount int
+	bufSize      int
+	// droppedDuration is the amount of time we should
+	// wait to check for dropped messages. Defaults
+	// to 3 seconds
+	droppedDuration time.Duration
+}
+
+// New creates a new Monitor. Start must be called in order to actually start
+// streaming logs
+func New(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) Monitor {
+	return new(buf, logger, opts)
+}
+
+func new(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) *monitor {
+	sw := &monitor{
+		logger:          logger,
+		logCh:           make(chan []byte, buf),
+		doneCh:          make(chan struct{}, 1),
+		bufSize:         buf,
+		droppedDuration: 3 * time.Second,
+	}
+
+	opts.Output = sw
+	sink := log.NewSinkAdapter(opts)
+	sw.sink = sink
+
+	return sw
+}
+
+// Stop deregisters the sink and stops the monitoring process
+func (d *monitor) Stop() {
+	d.logger.DeregisterSink(d.sink)
+	close(d.doneCh)
+}
+
+// Start registers a sink on the monitor's logger and starts sending
+// received log messages over the returned channel.
+func (d *monitor) Start() <-chan []byte {
+	// register our sink with the logger
+	d.logger.RegisterSink(d.sink)
+
+	streamCh := make(chan []byte, d.bufSize)
+
+	// run a go routine that listens for streamed
+	// log messages and sends them to streamCh
+	go func() {
```

Comment:
Why two goroutines instead of one here?

Reply:
See my above comment

Reply:
I think I would rework this into something like 

```
for {
  select1{}
  select2{}
}
```

where `select1` figures out if we should exit, warn about dropped messages, or consume a new message from `logCh`, and `select2` either exits or writes the warning or new message produced in `select1`.

Reply:
Ok, I'll give that a shot

Reply:
I managed to consolidate this down into 1 goroutine, 1 `for` loop, and 1 `select` statement that appears to work?  The tests still pass and it still seems to work fine when testing it from the console.  I'm interested to hear your thoughts.

Reply:
The reason I liked my idea of doing two selects is that it means we don't wind up trying to shove the dropped message into the logCh, of which we are the consumer.  It seems like with the current approach we may deadlock.

Reply:
Ok, I'll revisit this